pycernan/avro/client.py

```diff
@@ -1,47 +1,133 @@
 """
     Base Avro client from which all other clients derive.
 """
+import contextlib
 import socket
 
 import pycernan.avro.config
 
 from abc import ABCMeta, abstractmethod
+from queue import Queue, Empty
 
-from pycernan.avro.exceptions import EmptyBatchException
+from pycernan.avro.exceptions import EmptyBatchException, EmptyPoolException
 from pycernan.avro.serde import serialize
 
 
+DefunctConnection = None
+
+
+class TCPConnectionPool(object):
+    """
+    Fork friendly TCP connection pool.
+
+    Adapted from: https://github.com/gevent/gevent/blob/master/examples/psycopg2_pool.py
+    """
+    def __init__(self, host, port, maxsize, connect_timeout, read_timeout):
+        if maxsize <= 0:
+            raise ValueError("maxsize must be > 0")
+
+        self.host = host
+        self.port = port
+        self.size = 0
+        self.maxsize = maxsize
+
+        self.connect_timeout = connect_timeout
+        self.read_timeout = read_timeout
+        self.pool = None
+
+    def create_connection(self):
+        sock = socket.create_connection((self.host, self.port), timeout=self.connect_timeout)
+        sock.settimeout(self.read_timeout)
+        return sock
+
+    def get(self, _block=True):
+        if not self.pool:
+            # Lazy queue creation, ensuring we are fork() friendly.
+            # Caveat - Users must not use the connection pool prior to
+            # forking.
+            self.pool = Queue()
+
+        if self.size >= self.maxsize or self.pool.qsize():
```

Comment:
https://docs.python.org/3/library/queue.html#queue.Queue.qsize says:

> Note, qsize() > 0 doesn’t guarantee that a subsequent get() will not block

If I understand the logic here, the intent is to get an idle connection from the queue, if one exists. But given the statement above, that only works sometimes, and is subject to race conditions, etc, unless gevent somehow changes this to provide stronger guarantees?

Reply:
I understood it as to get an idle connection from the queue if there is one available and if not, then block until you get one. The `qsize` check encourages the reuse of connections until new ones are necessary, I think.

Reply:
Yes, the problem is just because the queue is not empty at that instant does not mean it won't be empty an instant later. The joys of multithreaded programming!

So say a client connects, then returns the connection to the pool. So there's 1 item in a queue. Now, two threads concurrently hit this block. `self.pool.qsize()` is 1 (actually the documentation only guarantees "approximately" 1) so both execute this conditional block, thinking they'll pop a connection out of the queue. One will, but the other one will be blocked, even though we _could_ just open a new connection.

Reply:
I think that's the behavior you want because that's what the first `self.size >= self.maxsize` check is enforcing. If we did make a new connection, we'd start violating the maxsize constraint and have an unbounded queue.

Reply:
that part works, but since the logic is something like:

```
if pool can't get bigger _or_ pool has an idle connection:
    ...
```

it can still happen that maybe the pool _could_ get bigger, but still more things go down this branch than there are idle connections in the pool.

Reply:
I adapted this code from an example which was assuming gevent usage.  In the gevent case, this logic works as intended as gevent semantics ensures we aren't falling into the same race conditions as with multithreading.

I will take a step back and re-attempt supporting multithreading.

Reply:
After looking at the [source code for gevents queue](http://www.gevent.org/_modules/gevent/queue.html#Queue.get), I'm 85% sure you are right. Still, [reasoning about this stuff is hard](https://glyph.twistedmatrix.com/2014/02/unyielding.html), so if it's not too hard, I'd approach any gevent code like real threads.

Otherwise maybe you want to use `gevent.queue` rather than relying on monkeypatching `queue`, just to make it very clear this code is correct _only_ with gevent.